### DISCUSSION TITLE
How does BAP know when all bpps have responded to search

### DISCUSSION THREAD

**Discussion by venkatramanm** [beckn/protocol-specifications#147] [Ideas]
After firing  search, BAP today doesnot know if all BPPs have responded or BAP needs to wait to collect more responses. 

Let's say gateway broadcasted search to 7 bpps and responses have come within 3 seconds from all BPP. TTL was 20 seconds. Why wait for 17 more seconds.? 🤔

**Comment by ravi-prakash-v** [comment]
Good point. There could be two approaches to solve this. 

**Solution 1:**
BG can send back the number of BPPs (n) it has broadcasted the Intent to as metadata the ACK response. The BAP will count the total number of unique responses and wait till it is equal to n or until the TTL expires whichever is sooner. 

**Solution 2:**
After it has forwarded the responses from the BPP the BPP can send an `on_search/end` response to indicate the end of the search cycle.

IMHO, I prefer Solution 1. It keeps the BG stateless. What do you think?

**Reply by venkatramanm** [reply]
Solution 1 is good. 

But this may delay the sending of ACK response  as it would need  to : 
1. do lookup, 
2. send the messages, 
3. receive acks from bpp, 
4. increment counter. 

Then send it to bap. ACK Was supposed to be quick. One alteration to this could be, 


1. BG sends ACK quickly
2. Then bg can call back special /on_search_meta  with meta on bpps count who sent ACK.

**Reply by venkatramanm** [reply]
@thejeshgn @pkapustin @core-wg-admin  

Any thoughts to take it forward?

**Reply by pkapustin** [reply]
I like solution 1 too. Gateway can be optimized to cache the number of BPPs in each domain / area, so I don't think it is a problem.

**Reply by thejeshgn** [reply]
I wait for only certain amount of time.  Currently I accept `on_search` responses for upto two minutes, responses after that I ignore. Because my search is also time sensitive. 

Overtime I will will have enough data to decide the timeout period.

**Reply by venkatramanm** [reply]
So the question remains is whether to
 1 alter ack response 
Or 
2. have a separate meta api with counts of BPP who sent ack.

**Reply by pkapustin** [reply]
I suggest to alter ACK response, as it is much simpler and reduces the chance of errors.

**Reply by thejeshgn** [reply]
Usually ACK will be returned by the input side QUEUE doing minimal non business validation. 

Any real response comes from the business system and that will be ASYNC. System may not be able to send you the number of BPPs successfully as part of SYNC ACK.

**Reply by pkapustin** [reply]
@thejeshgn Not sure about the other implementations, but our gateway is very simple and doesn't have any complexity like this. We can easily return the number of BPPs the request will be forwarded to at once. Personally I don't see why others shouldn't be able to do the same.

Also, please check #144 regarding sync APIs.
We already started using sync APIs for everything except search / on_search in the taxi network and are planning to use for the integrations in the public transport network.
It makes things much simpler for HTTP implementations and reduces possibilities for errors.

**Reply by pkapustin** [reply]
@thejeshgn I do see your point about queues though, in some cases it may not be feasible to return a response immediately (for example, on_search and for events that originate from the BPP).
But I don't think it is really the case with the number of BPPs the request is going to be forwarded to, I think it should be possible to return this at once.

**Reply by thejeshgn** [reply]
@pkapustin sure. But I think this is generally against the policy of ASYNC requests. May be @core-wg-admin  can clarify.

**Reply by pkapustin** [reply]
Yeah, essentially we thought that sync vs async could  be up to the network.
It can be quite beneficial to use sync for many APIs in HTTP implementations, see https://github.com/beckn/protocol-specifications/issues/144.

Of course, some things will have to stay async, like on_search and events that originate at the BPP.
